hydra-node/src/Hydra/Chain/Direct.hs

```diff
@@ -58,56 +158,280 @@ import Ouroboros.Network.Protocol.LocalTxSubmission.Server (
   localTxSubmissionServerPeer,
  )
 import qualified Ouroboros.Network.Protocol.LocalTxSubmission.Type as LocalTxSubmission
+import Ouroboros.Network.Server.RateLimiting (
+  AcceptedConnectionsLimit (..),
+ )
+import Ouroboros.Network.Socket (
+  SomeResponderApplication (..),
+  withServerNode,
+ )
 import qualified Shelley.Spec.Ledger.API as Ledger
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+
+data DirectChainLog
 
 withDirectChain ::
-  IO (Channel IO LByteString) ->
+  -- | Tracer for logging
   Tracer IO DirectChainLog ->
+  -- | Network configuration, defined by the genesis configuration.
+  --
+  -- See also 'defaultEpochSlots' and 'defaultNodeToClientVersionData' for playing
+  -- around.
+  (NodeToClientVersionData, EpochSlots) ->
+  -- | Socket used to connect to the server.
+  FilePath ->
   ChainComponent tx IO ()
-withDirectChain connect _tracer callback action = do
-  chan <- connect
-  action $ Chain{postTx = postTx chan}
+withDirectChain _tracer (vData, epochSlots) addr callback action = do
+  queue <- newTQueueIO
+  withIOManager $ \iocp -> do
+    race_
+      (action $ Chain{postTx = atomically . writeTQueue queue})
+      (connectTo (localSnocket iocp addr) tracers (versions queue) addr)
  where
-  postTx Channel{send} tx = do
-    -- TODO(SN): convert 'postChainTx' to a Cardano tx and send it to the node
-    send $ serialize ()
-    now <- getCurrentTime
-    callback (toOnChainTx now tx)
+  -- NOTE: written in such a way to make it easier to add support for multiple
+  -- versions if needed. A bit YAGNI but also tiny enough to be too much
+  -- overhead.
+  versions queue =
+    combineVersions
+      [ simpleSingletonVersions v vData (client (defaultCodecs epochSlots v) queue callback)
+      | v <- [nodeToClientVLatest]
+      ]
 
-data DirectChainLog
+  -- TODO: Provide tracers for these.
+  tracers :: NetworkConnectTracers LocalAddress NodeToClientVersion
+  tracers =
+    NetworkConnectTracers
+      { nctMuxTracer = nullTracer
+      , nctHandshakeTracer = nullTracer
+      }
+
+client ::
+  MonadSTM m =>
+  ClientCodecs Block m ->
+  TQueue m (PostChainTx tx) ->
+  ChainCallback tx m ->
+  OuroborosApplication 'InitiatorMode LocalAddress LByteString m () Void
+client codecs queue callback =
+  OuroborosApplication $ \_connectionId _controlMessageSTM ->
+    [ localChainSyncMiniProtocol
+    , localTxSubmissionMiniProtocol
+    ]
+ where
+  -- TODO: Factor out the tracer work on Hydra.Network.Ouroboros and use it to
+  -- provide SendRecv traces for both protocols.
+  localChainSyncMiniProtocol =
+    MiniProtocol
+      { miniProtocolNum = MiniProtocolNum 5
+      , miniProtocolLimits = maximumMiniProtocolLimits
+      , miniProtocolRun = InitiatorProtocolOnly initiator
+      }
```

Comment:
Is this config not re-usable from `ouroboros-network`? Would be nice to not hard-code the protocol numbers at least.